File: ios/Sources/WebviewPluginPlugin/WebviewPluginPlugin.swift
import Foundation
import Capacitor
@preconcurrency import WebKit

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(WebviewPluginPlugin)
public class WebviewPluginPlugin: CAPPlugin, CAPBridgedPlugin {
  var webViewController: UIViewController?
  public let identifier = "WebviewPluginPlugin"
  public let jsName = "WebviewPlugin"
  private static let productionUrl = "https://webpayment.moneytrack.io"
  private static let stagingUrl = "https://webpayment-staging.moneytrack.io"
  
  enum Environment: String {
    case production = "production"
    case staging = "staging"
  }
  var apiUrl: String = productionUrl
  public let pluginMethods: [CAPPluginMethod] = [
    CAPPluginMethod(name: "setEnvironment", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "openAddCard", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "openPay", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "openMap", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "openOnboarding", returnType: CAPPluginReturnPromise)
  ]
  
  @objc func setEnvironment(_ call: CAPPluginCall) {
    guard let envString = call.getString("environment"),
          let environment = Environment(rawValue: envString.lowercased()) else {
      call.reject("Invalid environment. Use 'production' or 'staging'.")
      return
    }

    switch environment {
    case .production:
      apiUrl = WebviewPluginPlugin.productionUrl
    case .staging:
      apiUrl = WebviewPluginPlugin.stagingUrl
    }

    call.resolve()
  }

  @objc func openAddCard(_ call: CAPPluginCall) {
    openWebView(url: "\(apiUrl)/#/add-card", call: call)
  }

  @objc func openPay(_ call: CAPPluginCall) {
    openWebView(url: "\(apiUrl)/#/pay", call: call)
  }

  @objc func openMap(_ call: CAPPluginCall) {
    openWebView(url: "\(apiUrl)/#/map", call: call)
  }

  @objc func openOnboarding(_ call: CAPPluginCall) {
    guard let payload = call.getString("payload") else {
      call.reject("Payload is required.")
      return
    }
    openWebView(
      url: "\(apiUrl)/#/onboarding?payload=\(payload)",
      call: call
    )
  }

  private func openWebView(url: String, call: CAPPluginCall) {
    DispatchQueue.main.async {
      let config = WKWebViewConfiguration()
      self.webView = WKWebView(
        frame: UIScreen.main.bounds,
        configuration: config
      )
      self.webView?.navigationDelegate = self
              
      let urlRequest = URLRequest(url: URL(string: url)!)
      self.webView?.load(urlRequest)

      let vc = UIViewController()
      vc.view = self.webView
      self.webViewController = vc

      self.bridge?.viewController?.present(vc, animated: true, completion: nil)
    }
  }

}

extension WebviewPluginPlugin: WKNavigationDelegate {
    
  public func webView(
    _ webView: WKWebView,
    decidePolicyFor navigationAction: WKNavigationAction,
    decisionHandler: @escaping (WKNavigationActionPolicy) -> Void
  ) {
    if let url = navigationAction.request.url?.absoluteString, url
      .contains("sdk-static-page") {
      DispatchQueue.main.async {
        self.webViewController?.dismiss(animated: true, completion: nil)
        self.webView = nil
        self.webViewController = nil
        self.notifyListeners("dismissed", data: ["dismissed": true])
      }
    }
    decisionHandler(.allow)
  }
}

File: android/src/main/java/com/moneytrack/moneytrackpay/WebviewPluginPlugin.java
package com.moneytrack.moneytrackpay;

import android.content.Context;
import android.view.ViewGroup;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "WebviewPlugin")
public class WebviewPluginPlugin extends Plugin {
    public static final String productionUrl = "https://webpayment.moneytrack.io";
    public static final String stagingUrl = "https://webpayment.jestaging.moneytrack.io";
    private String apiUrl = productionUrl;

    @PluginMethod
    public void setEnvironment(PluginCall call) {
        String environment = call.getString("environment");
        if (environment == null) {
            call.reject("Environment is required.");
            return;
        }

        switch (environment.toLowerCase()) {
            case "production":
                apiUrl = productionUrl;
                break;
            case "staging":
                apiUrl = stagingUrl;
                break;
            default:
                call.reject("Invalid environment. Use 'production' or 'staging'.");
                return;
        }

        call.resolve();
    }

    private WebView webView;

    @PluginMethod
    public void openAddCard(PluginCall call) {
        openWebView(apiUrl + "/#/add-card", call);
    }

    @PluginMethod
    public void openPay(PluginCall call) {
        openWebView(apiUrl + "/#/pay", call);
    }

    @PluginMethod
    public void openMap(PluginCall call) {
        openWebView(apiUrl + "/#/map", call);
    }

    @PluginMethod
    public void openOnboarding(PluginCall call) {
        String payload = call.getString("payload", "");
        openWebView(apiUrl + "/#/onboarding?payload=" + payload, call);
    }

    private void openWebView(String url, PluginCall call) {
        getActivity().runOnUiThread(() -> {
            Context context = getContext();

            // Create WebView
            webView = new WebView(context);
            WebSettings webSettings = webView.getSettings();
            webSettings.setJavaScriptEnabled(true);
            webSettings.setDomStorageEnabled(true);
            webSettings.setAllowFileAccess(true);
            webSettings.setAllowContentAccess(true);

            webView.setLayoutParams(new ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.MATCH_PARENT
            ));

            // Set WebViewClient
            webView.setWebViewClient(new WebViewClient() {
                @Override
                public boolean shouldOverrideUrlLoading(WebView view, String url) {
                    if (url.contains("static-page")) {
                        // Remove WebView when dismissed
                        ((ViewGroup) webView.getParent()).removeView(webView);
                        call.resolve(new JSObject().put("dismissed", true));
                        return true;
                    }
                    return super.shouldOverrideUrlLoading(view, url);
                }
            });

            // Load the URL
            webView.loadUrl(url);

            // Get the root view and add WebView
            ViewGroup rootView = (ViewGroup) getActivity().findViewById(android.R.id.content);
            rootView.addView(webView);
        });
    }
}
